SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,956 @@
+// OTHER DEALINGS IN THE SOFTWARE.
+
+// Using `const`s instead of `static`s for performance
+const SPIFLASHLOGGER_SECTOR_SIZE = 4096; // Size of sectors
+const SPIFLASHLOGGER_SECTOR_METADATA_SIZE = 3; // Size of metadata at start of sectors
+const SPIFLASHLOGGER_SECTOR_BODY_SIZE = 4093; // Size of writeable memory / sector
+const SPIFLASHLOGGER_CHUNK_SIZE = 256; // Number of bytes we write / operation
+
+const SPIFLASHLOGGER_OBJECT_MARKER = "\x00\xAA\xCC\x55";
+const SPIFLASHLOGGER_OBJECT_MARKER_SIZE = 4;
+
+const SPIFLASHLOGGER_OBJECT_HDR_SIZE = 7; // SPIFLASHLOGGER_OBJECT_MARKER (4 bytes) + size (2 bytes) + crc (1 byte)
+const SPIFLASHLOGGER_OBJECT_MIN_SIZE = 6; // SPIFLASHLOGGER_OBJECT_MARKER (4 bytes) + size (2 bytes)
+
+// Default status of sector is free
+// which is equal to the 0xFFFFFF
+const SECTOR_STATUS_FREE = 0;
+// Indicates that we can read metadata from that sector
+const SECTOR_HAS_METADATA = 1;
+// Indicates that sector has some written data
+const SECTOR_WRITE_START = 0x02;
+// Indicates that sector was closed on writing
+const SECTOR_WRITE_DONE = 0x04;
+// Indicates that sector has at least one start code
+const SECTOR_HAS_START_CODE = 0x08;
+// indicates that start code could be invalid
+// because object was erased
+const SECTOR_HAS_REMOVED_OBJECT = 0x08;
+
+class SPIFlashLogger {
+
+    static version = "3.0.0";
+
+    _flash = null; // hardware.spiflash or an object with an equivalent interface
+    _serializer = null; // github.com/electricimp/serializer (or an object with an equivalent interface)
+
+    _size = null; // The size of the spiflash
+    _start = null; // First block to use for logging
+    _end = null; // Last block to use for logging
+    _len = null; // The length of the flash available (end-start)
+    _sectors = 0; // The number of sectors in _len
+    _max_data = 0; // The maximum data we can push at once
+
+    _at_sec = 0; // Current sector we're writing to
+    _at_pos = 0; // Current position we're writing to in the sector
+
+    _enables = 0; // Counting semaphore for _enable/_disable
+    _sectorsMap = null; // list of sector objects
+
+    // This calss uses to read the serialized object from the flash
+    // and de-serialize it.
+    _loggerSerializedObject = class {
+        _flash = null;
+        _logger = null;
+        _addr = null;
+        _isValid = false;
+        _payload = null;
+        _len = 0;
+
+        // constructor
+        // Parameters:
+        //      addr - object address
+        //      logger - the SPIFlashLogger object
+        //      flash - the spiflash object
+        constructor(addr, logger, flash) {
+            _logger = logger;
+            _flash = flash;
+            _addr = addr;
+
+            _isValid = false;
+            _payload = null;
+            _len = 0;
+
+            init(addr);
+        }
+
+        // internal method to initialize object:
+        // check start code and read object size
+        function init(pos) {
```

Comment:
if it's internal only it should be prefixed with `_`.

also why do we need this method at all? It's never called from anywhere else.

Reply:
joined to constructor, thanks